core/controllers/suggestion_test.py

```diff
@@ -0,0 +1,220 @@
+                        'old_value': self.old_content,
+                        'new_value': self.new_content
+                    },
+                    'description': 'change to state 2',
+                    'final_reviewer_id': self.reviewer_id,
+                    'assigned_reviewer_id': self.assigned_reviewer_id
+                }, csrf_token)
+
+            self.post_json(
+                '%s/' % feconf.SUGGESTION_URL_PREFIX, {
+                    'suggestion_type': (
+                        suggestion_models.SUGGESTION_TYPE_EDIT_STATE_CONTENT),
+                    'target_type': suggestion_models.TARGET_TYPE_EXPLORATION,
+                    'target_id': 'exp1',
+                    'target_version_at_submission': exploration.version,
+                    'change_cmd': {
+                        'cmd': exp_domain.CMD_EDIT_STATE_PROPERTY,
+                        'property_name': exp_domain.STATE_PROPERTY_CONTENT,
+                        'state_name': 'State 3',
+                        'old_value': self.old_content,
+                        'new_value': self.new_content
+                    },
+                    'description': 'change to state 3',
+                    'final_reviewer_id': self.reviewer_id,
+                    'assigned_reviewer_id': self.assigned_reviewer_id
+                }, csrf_token)
+            self.logout()
+
+    def test_create_suggestion(self):
+        self.login(self.AUTHOR_EMAIL_2)
+        if constants.USE_NEW_SUGGESTION_FRAMEWORK:
+
+            response = self.testapp.get('/explore/%s' % self.EXP_ID)
+            csrf_token = self.get_csrf_token_from_response(response)
+
+            exploration = exp_services.get_exploration_by_id(self.EXP_ID)
+            self.post_json(
+                '%s/' % feconf.SUGGESTION_URL_PREFIX, {
+                    'suggestion_type': (
+                        suggestion_models.SUGGESTION_TYPE_EDIT_STATE_CONTENT),
+                    'target_type': suggestion_models.TARGET_TYPE_EXPLORATION,
+                    'target_id': 'exp1',
+                    'target_version_at_submission': exploration.version,
+                    'change_cmd': {
+                        'cmd': exp_domain.CMD_EDIT_STATE_PROPERTY,
+                        'property_name': exp_domain.STATE_PROPERTY_CONTENT,
+                        'state_name': 'State 3',
+                        'old_value': self.old_content,
+                        'new_value': self.new_content
+                    },
+                    'description': 'change again to state 3',
+                }, csrf_token)
+            suggestions = self.get_json(
+                '%s?list_type=author&author_id=%s' % (
+                    feconf.SUGGESTION_LIST_URL_PREFIX,
+                    self.author_id_2))['suggestions']
+            self.assertEqual(len(suggestions), 3)
+        self.logout()
+
+    def test_accept_suggestion(self):
+        if constants.USE_NEW_SUGGESTION_FRAMEWORK:
```

Comment:
ditto: when will this test actually run? Testing locally is not enough. Perhaps you should set USE_NEW_SUGGESTION_FRAMEWORK to True within the context and test this?

Reply:
I tried switching it to true. but the issue is the fact that the handlers are assigned in main.py (which runs before this test) and hence just setting it to true doesn't work as expected. Another thing I can do is use different URL prefixes (probably like generalsuggestionhandler in place of suggestionhandler) and then not use the switch in the backend. i.e. both handlers will be active in the backend, and we will switch in the frontend. This way we can run the tests too. Does this sound ok?

Reply:
Talk to @aks681, he is doing similar stuff with the new structures project (where there's functionality that is not enabled but is still testable). See if you can reuse some of those patterns.

Reply:
I took a look at how the ENABLE_NEW_STRUCTURES is implemented. I believe I am doing it wrong, but the situation is slightly different here. So if the flag is False, I want to route to the old handlers and if it is true I want to route to the new handlers. The old and new handlers have different URL params also. I think it would make sense to have new URLs and return PageNotFound if the flag is False (similar to the ENABLE_NEW_STRUCTURES implementation). Shall I do this?

Reply:
Yes, that sounds fine. Thanks!

Reply:
Done